### PR TITLE
docs: add Supporter Events Service details

### LINKED ISSUE #18: Add Supporter Events Details
Document the purpose of Supporter Events and how reads are going to be handled.

### PR DESCRIPTION
**Pull request by dipasqualew**
Document the purpose of Supporter Events and how reads are going to be handled.

Closes #18

### CODE REVIEW

**Comment by corinja** [comment]
This is a good start. Can we have some definitions or links to help explain the first line - answering the question "what are `supporter data` and `supporter events`"? 

Maybe `Supporter Data` links to the `data-models` or `data-contracts` repos? `Supporter Events` I guess is things like donating, paying in fundraising money, making a giftaid declaration - it would be good to just have a brief definition of the kind of events we're talking about.

File: docs/architecture.md [hunk 1]
```diff
@@ -3,7 +3,24 @@
...
+
+The service aims to propagate supporter data to User Service and supporter events to ERP Service.
+
+User Service and ERP Service exist on one region only because of their structure; should they be unavailable for any reason, propagation of data from upstream services may be interrupted. Supporter Events is a multi-region service designed to handle incoming traffic from all sources. Events are queued to SQS and processed one by one.
+
+When identities are supplied to the `user-service` section of the payload, Supporter Events returns a salted hash of the identity that can be used by upstream services to identity resources, so that anonymisation won't break relationships between data and services.
+
+Supporter Service is supposed to handle only write requests to User Service and ERP Service. If an upstream service requires data stored in either service, it should either:
+
+- **For critical paths**: store a copy of the data in the service itself
```
Comment (corinja) [inline docs/architecture.md:16]: Are the `paths` defined somewhere? E.g. what is on the `critical` and what is on the `non-critical` path?

File: docs/architecture.md [hunk 2]
```diff
@@ -3,7 +3,24 @@
...
+- **For non-critical paths**: gracefully handle request failures
+
+Every service should be able to operate independently from User Service and ERP Service for all the critical paths; when a path is deemed non critical (i.e. it is acceptable to present a failure to an end user), User Service and ERP Service can be queued directly.
+
+While User Service and ERP Service are unlikely to be unavailable, under no circumstances a critical path upstream should depend on either, for read or writes, as they are not multiregional services.
```
Comment (corinja) [inline docs/architecture.md]: Just a grammatical thing here. It should read "under no circumstances should a critical path upstream depend on either, for read or for writes, as they are not multiregional services".

File: docs/architecture.md (hunk 1 above)
Comment (dipasqualew) [reply docs/architecture.md:16]: Not at the moment

File: docs/architecture.md (hunk 2 above)
Comment (dipasqualew) [reply docs/architecture.md]: Thanks Corin! 14e9157906d1f22d6914edb17c2078cb3c11bff1

**Comment by dipasqualew** [comment]
I think details about the expected payload and the response should be defined in the Supporter Events repo documentation, otherwise this repo will blob and be hard to keep up to date.

**Review by corinja — APPROVED** [review]

File: docs/architecture.md [hunk 3]
```diff
@@ -3,7 +3,24 @@
...
-## Supporter data flow
+## Supporter Events Service
+
+The service aims to propagate [supporter identities](https://github.com/comicrelief/data-models/tree/master/src/schemas/user-service/models/requests) to User Service and [supporter events](https://github.com/comicrelief/data-models/tree/master/src/schemas/erp/models/actions) to ERP Service.
+
+User Service and ERP Service exist on one region only because of their structure; should they be unavailable for any reason, propagation of data from upstream services may be interrupted. Supporter Events is a multi-region service designed to handle incoming traffic from all sources. Events are queued to SQS and processed one by one.
+
+When identities are supplied to the `user-service` section of the payload, Supporter Events returns a salted hash of the identity that can be used by upstream services to identity resources, so that anonymisation won't break relationships between data and services.
+
+Supporter Service is supposed to handle only write requests to User Service and ERP Service. If an upstream service requires data stored in either service, it should either:
```
Comment (seb-cr) [inline docs/architecture.md]: Do you have a preferred naming style for services? I'm happy with Title Case service names that include the word "Service". There are a couple of deviations here though: "Supporter Events" and "Supporter Service" should both be "Supporter Event Service" by this convention.

Also, the repo name uses "Event" in singular, which I think sounds best (just as we have a User Service rather than a Users Service). I noticed I made this mistake too in the diagram... shall I push a commit to fix that, or are you happy to include it in one of your fixups?

File: docs/architecture.md [hunk 4]
```diff
@@ -3,7 +3,24 @@
...
+
+- **For critical paths**: store a copy of the data in the service itself
+- **For non-critical paths**: gracefully handle request failures
+
+Every service should be able to operate independently from User Service and ERP Service for all the critical paths; when a path is deemed non critical (i.e. it is acceptable to present a failure to an end user), User Service and ERP Service can be queued directly.
```
Comment (seb-cr) [inline docs/architecture.md]: I'm not sure about "User Service and ERP Service can be queued directly" – do you mean that requests can be made directly to these services instead of going via the Supporter Event Service?

Also, very minor point, could you hyphenate "non critical" please?

**Review by seb-cr — COMMENTED** [review]
Thanks for making a start on this!

File: docs/architecture.md (hunk 1 above)
Comment (seb-cr) [reply docs/architecture.md:16]: There's a definition given just below saying that a non-critical path is one where "it is acceptable to present a failure to an end user". Would it work to move that up so that it's next to the first mention of critical/non-critical paths?

File: docs/architecture.md (hunk 3 above)
Comment (dipasqualew) [reply docs/architecture.md]: I will fix it, cheers! 🙂

Comment (dipasqualew) [reply docs/architecture.md]: re: preferred names. I would vote for backticked repo names, i.e. `serverless-user-service` and `serverless-erp-service`. A bit verbose, but this way we know exactly what we are talking about.

Comment (seb-cr) [reply docs/architecture.md]: Thanks!

File: docs/architecture.md (hunk 4 above)
Comment (dipasqualew) [reply docs/architecture.md]: Yes you can query those services directly, if you have the API Key. Nothing stops you from doing so if you need an immediate answer. Supporter Event Service only handles writes, we will need to read the data as well

File: docs/architecture.md (hunk 1 above)
Comment (dipasqualew) [reply docs/architecture.md:16]: 👍🏻

Comment (dipasqualew) [reply docs/architecture.md:16]: 4ffd216817f068c20575913aab8f8dac65cee19f

File: docs/architecture.md (hunk 3 above)
Comment (dipasqualew) [reply docs/architecture.md]: 858352228cbf9e2e6cd3cfdc7fd381cd164ccf73

File: docs/architecture.md (hunk 4 above)
Comment (seb-cr) [reply docs/architecture.md]: Ah, I think it might have been a typo – it makes sense if you replace "queued" with "queried"!

File: docs/architecture.md [hunk 5]
```diff
@@ -3,46 +3,65 @@
...
+`serverless-user-service` and `serverless-erp-service` exist on one region only because of their structure; should they be unavailable for any reason, propagation of data from upstream services may be interrupted. `supporter-event-service` is a multi-region service designed to handle incoming traffic from all sources. Events are queued to SQS and processed one by one.
+
+When identities are supplied to the `serverless-user-service` section of the payload, `supporter-event-service` returns a salted hash of the identity that can be used by upstream services to identity resources, so that anonymisation won't break relationships between data and services.
+
+`supporter-event-service` is supposed to handle only write requests to `serverless-user-service` and `serverless-erp-service`. If an upstream service requires data stored in either service, it should either:
+
+- **For critical paths**: store a copy of the data in the service itself
+- **For non-critical paths**: gracefully handle request failures
+
+Critical path here means whether it is acceptable or not to present a failure to and end user.
```
Comment (seb-cr) [inline docs/architecture.md]: Small typo: "and end user" -> "an end user".

Perhaps also rephrase this to explitly state what a critical path is: `A *critical path* is one where it is not acceptable to present a failure to an end user.`

Comment (dipasqualew) [reply docs/architecture.md]: 20a4a2bfb5271211e287ae50ed86cd6393710a28

File: docs/architecture.md (hunk 4 above)
Comment (dipasqualew) [reply docs/architecture.md]: 329dc9b7d3f2cb0741de8bfb16da812f8ce648c4

**Review by seb-cr — APPROVED** [review]

**Review by lebaz20 — APPROVED** [review]
